crypto/bio/bss_dgram.c

```diff
@@ -294,26 +308,262 @@ static void dgram_reset_rcv_timeout(BIO *b)
 # endif
 }
 
+# if defined(HAVE_IP_PKTINFO)
+/* LINUX has IP_PKTINFO for IPv4 */
+static int dgram_read_unconnected_v4(BIO *b, char *in, int inl,
+                                     int flags,
+                                     BIO_ADDR *dstaddr, BIO_ADDR *peer)
+{
+    int len = 0;
+    unsigned char chdr[CMSG_SPACE(sizeof(struct in_pktinfo))];
+    struct iovec iov;
+    struct msghdr mhdr;
+    struct in_pktinfo *pkt_info;
+    struct cmsghdr *cmsg;
+    int val;
+
+    /* enable PKTINFO receive */
+    val = 1;
+    if (setsockopt(b->num, IPPROTO_IP, IP_PKTINFO, &val, sizeof(val)) < 0) {
+        return -1;
+    }
+
+    memset(&iov, 0, sizeof(iov));
+    iov.iov_len = inl;
+    iov.iov_base = (caddr_t) in;
+
+    memset(&mhdr, 0, sizeof(mhdr));
+    mhdr.msg_name = (caddr_t) BIO_ADDR_sockaddr(peer);
+    mhdr.msg_namelen = sizeof(struct sockaddr_in);
+    mhdr.msg_iov = &iov;
+    mhdr.msg_iovlen = 1;
+
+    if (dstaddr) {
+        memset(chdr, 0, sizeof(chdr));
+        cmsg = (struct cmsghdr *)chdr;
+        mhdr.msg_control = (void *)cmsg;
+        mhdr.msg_controllen = sizeof(chdr);
+    }
+
+    pkt_info = NULL;
+    if ((len = recvmsg(b->num, &mhdr, 0)) >= 0) {
+        for (cmsg = CMSG_FIRSTHDR(&mhdr);
+             cmsg != NULL; cmsg = CMSG_NXTHDR(&mhdr, cmsg)) {
+            if (cmsg->cmsg_level != IPPROTO_IP)
+                continue;
+            switch (cmsg->cmsg_type) {
+            case IP_PKTINFO:
+                pkt_info = (struct in_pktinfo *)CMSG_DATA(cmsg);
+                break;
+            }
+        }
+
+        /* see if we found something */
+        if (pkt_info != NULL && dstaddr != NULL) {
+            dstaddr->s_in.sin_family = AF_INET;
+            dstaddr->s_in.sin_addr = pkt_info->ipi_addr;
+        }
+    }
+
+    /* NOTE: peer was filled in by kernel */
+    return len;
+}
+# elif defined(HAVE_IP_RECVDSTADDR)
+/* FREEBSD, DragonFly, NetBSD, OpenBSD, probably BSDi */
+/*    implies IP_SENDSRCADDR is available too         */
```

Comment:
Similarly here it would probably be better to use `__FreeBSD__`, `__NetBSD__` etc, rather than add a new define to Configure. Probably we could add some pre-processor logic to the beginning of the file to do all the system specific macro checks and then just define `HAVE_IP_RECVDSTADDR` or `HAVE_IP_PKTINFO` just for the scope of this file.

Reply:
It seemed like the rest of the networking is based upon capabilities rather than operating system defines. I put the HAVE_IP_RECVDSTADDR setting into Configure in the generic BSD section.  I can do it as you suggest, it just seemed to make sense to put it in central spot.